### ISSUE TITLE
Unable to upload symbols

### ISSUE THREAD

**Issue by ramx** [loic-sharma/BaGet#245]
### Describe the bug
I'm not able to upload symbol file using nuget

C:\Program Files (x86)\Jenkins\workspace\libLogger>c:\nuget\nuget.exe push "C:\Program Files (x86)\Jenkins\workspace\libLogger\bin\Debug\Logger.*.snupkg" 1234567890 -Source http://192.168.21.42:5000/v3/index.json 
Pushing Logger.2.0.0.snupkg to 'http://192.168.21.42:5000/api/v2/symbol'...
  PUT http://192.168.21.42:5000/api/v2/symbol/
  BadRequest http://192.168.21.42:5000/api/v2/symbol/ 140ms
Response status code does not indicate success: 400 (Bad Request)

**Comment by loic-sharma** [comment]
Thanks for filing this bug, I’ll try to reproduce this! What version of NuGet are you using? Is there anything special about your project’s snupkg? Does BaGet’s logs indicate anything out of the ordinary, like an exception?

**Comment by fuzzybair** [comment]
I am seeing the same error with symbol uploads using version 0.1.70 and 0.1.71 however when running BaGet built from master within VS Code to debug I do not get an error. git diff tells me there is no difference between the tag for 0.1.71 and the current master. I am not sure where to go with that.

```
.NET Core SDK (reflecting any global.json):
 Version:   2.2.105
 Commit:    7cecb35b92

Runtime Environment:
 OS Name:     centos
 OS Version:  7
 OS Platform: Linux
 RID:         centos.7-x64
 Base Path:   /usr/share/dotnet/sdk/2.2.105/

Host (useful for support):
  Version: 2.2.3
  Commit:  6b8ad509b6

.NET Core SDKs installed:
  2.2.105 [/usr/share/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.2.3 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.2.3 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.2.3 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
```

v0.1.71
```
 ~/NuGetPackages/Deploy $ dotnet nuget push -s http://localhost:5000/v3/index.json MTK-Logger-devel.1.4.0.2816.nupkg 
warn : No API Key was provided and no API Key could be found for 'http://localhost:5000/api/v2/package'. To save an API Key for a source use the 'setApiKey' command.
info : Pushing MTK-Logger-devel.1.4.0.2816.nupkg to 'http://localhost:5000/api/v2/package'...
info :   PUT http://localhost:5000/api/v2/package/
info :   Created http://localhost:5000/api/v2/package/ 1008ms
info : Your package was pushed.
warn : Found symbols package 'MTK-Logger-devel.1.4.0.2816.snupkg', but no API key was specified for the symbol server. To save an API Key, run 'NuGet.exe setApiKey [your API key from http://www.NuGet.org]'.
warn : No API Key was provided and no API Key could be found for 'http://localhost:5000/api/v2/symbol'. To save an API Key for a source use the 'setApiKey' command.
info : Pushing MTK-Logger-devel.1.4.0.2816.snupkg to 'http://localhost:5000/api/v2/symbol'...
info :   PUT http://localhost:5000/api/v2/symbol/
info :   BadRequest http://localhost:5000/api/v2/symbol/ 57ms
error: Response status code does not indicate success: 400 (Bad Request).
```
Debug build
```
 ~/NuGetPackages/Deploy $ dotnet nuget push -s http://localhost:50561/v3/index.json MTK-Logger-devel.1.4.0.2816.nupkg 
warn : No API Key was provided and no API Key could be found for 'http://localhost:50561/api/v2/package'. To save an API Key for a source use the 'setApiKey' command.
info : Pushing MTK-Logger-devel.1.4.0.2816.nupkg to 'http://localhost:50561/api/v2/package'...
info :   PUT http://localhost:50561/api/v2/package/
info :   Created http://localhost:50561/api/v2/package/ 1164ms
info : Your package was pushed.
warn : Found symbols package 'MTK-Logger-devel.1.4.0.2816.snupkg', but no API key was specified for the symbol server. To save an API Key, run 'NuGet.exe setApiKey [your API key from http://www.NuGet.org]'.
warn : No API Key was provided and no API Key could be found for 'http://localhost:50561/api/v2/symbol'. To save an API Key for a source use the 'setApiKey' command.
info : Pushing MTK-Logger-devel.1.4.0.2816.snupkg to 'http://localhost:50561/api/v2/symbol'...
info :   PUT http://localhost:50561/api/v2/symbol/
info :   Created http://localhost:50561/api/v2/symbol/ 319ms
info : Your package was pushed.
```

**Comment by loic-sharma** [comment]
Thanks for the report @fuzzybair. I installed BaGet v0.1.71 and .NET Core 2.2.105 and I can't reproduce this issue. Some more questions:

1. What storage type are you using?
2. What database type are you using?
3. Is there anything special about your snupkg? Would you mind attaching a nuget package and its snupkg here that reproduces this issue?
4. How are you running BaGet? Is it `dotnet BaGet.dll` or through the docker image?
4. If you run BaGet on another operating system, does this issue still reproduce?

**Comment by fuzzybair** [comment]
1. FileSystem
2. Sqlite
3. NuGet and symbol package built on a Jenkins build server(windows). using "nuget.exe pack Logger.csproj -Properties "Platform=x64;tag=-devel" -Symbols -SymbolPackageFormat snupkg" contents of the snupkg is just a pdb file set to Portable 
![image](https://user-images.githubusercontent.com/5714774/55344560-e6fb3480-546a-11e9-8036-fcda5b64a0ad.png)

I can not release our code but I will work on getting a test package that reproduces the error.

4. I am running "dotnet BaGet.dll" via systemd.
BaGet.service
````
[Unit]
Description=A lightweight NuGet and Symbol server.

[Service]
ExecStart=/usr/bin/dotnet BaGet.dll
WorkingDirectory=/opt/BaGet/server
EnvironmentFile=/opt/BaGet/conf/BaGet.env
PassEnvironment=ASPNETCORE_URLS
Restart=always
RestartSec=120
SyslogIdentifier=BaGet

[Install]
WantedBy=multi-user.target
````
BaGet.env
````
ASPNETCORE_URLS=http://*:5000
````

appsettings.json
````
{
  "ApiKey": "", 
  "PackageDeletionBehavior": "Unlist",
  "AllowPackageOverwrites": false,

  "Database": {
    "Type": "Sqlite",
    "ConnectionString": "Data Source=/opt/BaGet/data/baget.db"
  },  

  "Storage": {
    "Type": "FileSystem",
    "Path": "/opt/BaGet/data"
  },  

  "Search": {
    "Type": "Database"
  },  

  "Mirror": {
    "Enabled": false,
    "PackageSource": "https://api.nuget.org/v3/index.json"
  },  

  "Logging": {
    "IncludeScopes": false,
    "Debug": {
      "LogLevel": {
        "Default": "Warning"
      }   
    },  
    "Console": {
      "LogLevel": {
        "Default": "Warning"
      }   
    }   
  }
}
````

5. tessting on windows 10 with dontnet 2.2.103 and version 0.1.71(default configuration) using the same nupkg/snupkg does not seen to have the same issue.

**Comment by fuzzybair** [comment]
I am not opposed to switching to the Docker based approach. However I have not had much luck in the past getting docker containers running as Linux that will come up after system reboot. If you could provide a link or some documentation on how to get BaGet running under systemd with a docker, I am willing to try it.

**Comment by 7thcubic** [comment]
@fuzzybair https://docs.docker.com/engine/reference/run/#restart-policies---restart

**Comment by fuzzybair** [comment]
@7thcubic Thanks for the link, with it I was able to get the docker instance working and restarting after reboot.  As a note after setting the docker services to restart on boot, two changes to the docker run command were needed for it to work 

1. Replace --rm with  --restart=unless-stopped
2. Add a :Z to the end of the volume mount (for SeLinux context)

```
docker run --restart=unless-stopped --name nuget-server -p 5555:80 --env-file /opt/BaGet/baget.env -v "/opt/BaGet:/var/baget:Z" loicsharma/baget:latest
```

with the docker version I am able to upload symbols. I am not sure why running under systemd directly did not work.

**Comment by fuzzybair** [comment]
I was able to upload but Visual Studio did not find the symbols I believe the cause is Issue #224

**Comment by loic-sharma** [comment]
Hey @fuzzybair, I wasn't able to reproduce the issue in https://github.com/loic-sharma/BaGet/issues/224. I created a project, targeted the x64 platform, packed it using nuget.exe, and loaded the symbols using Visual Studio 2017. This did not reproduce the issue. Could you provide a sample project that when packed reproduces this issue?

**Comment by loic-sharma** [comment]
@MatijaSestak Could you provide a sample project that when packed reproduces this issue?

**Comment by MatijaSestak** [comment]
Sure, https://github.com/MatijaSestak/AppLogger

**Comment by bhattvishal** [comment]
I am having the same issue as mentioned above. Here are my development environment

- Visual Studio 2017

- .Net Framework Project (4.6.2)

- NuGet.exe (v5.2.0)

I am not sure, but this issue could be because BaGet uses Nuget v2 api for symbols for snupkg packages. But, as stated [here ](https://docs.microsoft.com/en-us/nuget/create-packages/symbol-packages-snupkg)it is only supported in NuGet api v3 for Symbols

> If the symbol package isn't published, check that you've configured the NuGet.org source as https://api.nuget.org/v3/index.json. Symbol package publishing is only supported by the NuGet V3 API.

**Comment by loic-sharma** [comment]
> Sure, https://github.com/MatijaSestak/AppLogger

@MatijaSestak [This line ](https://github.com/MatijaSestak/AppLogger/blob/4c30e5f257f987c8b918dcb68c3669ab3f8fe156/AppLogger/AppLogger.csproj#L20) looks suspicious, you should use `<DebugType>portable</DebugType>` in both debug and release builds (it looks like your release builds should work though, see [this line](https://github.com/MatijaSestak/AppLogger/blob/4c30e5f257f987c8b918dcb68c3669ab3f8fe156/AppLogger/AppLogger.csproj#L28))

@bhattvishal @SiOuz Please make sure you don't have `<DebugType>full</DebugType>` or `<DebugType>pdbonly</DebugType>` anywhere in your .csproj files. You should use `<DebugType>portable</DebugType>` instead as BaGet does not support Windows PDBs yet. See: https://github.com/loic-sharma/BaGet/issues/153

> I am not sure, but this issue could be because BaGet uses Nuget v2 api for symbols for snupkg packages. But, as stated here it is only supported in NuGet api v3 for Symbols

TLDR: BaGet is doing the right thing here. More detailed explanation: the V3 API uses the V2 API for package uploads. If you go to [nuget.org's V3 package source](https://api.nuget.org/v3/index.json), you'll find the following JSON:

```
{
  "version": "3.0.0",
  "resources": [
    ...

    {
      "@id": "https://www.nuget.org/api/v2/package",
      "@type": "PackagePublish/2.0.0"
    },

    ...
  ]
}
```

That being said, you MUST use a V3 package source URL (like `https://api.nuget.org/v3/index.json`) instead of a V2 package source URL (like `https://nuget.org/`) to publish symbol packages. The NuGet protocol can be awfully confusing 😅

**Comment by trgiangvp3** [comment]
> That being said, you MUST use a V3 package source URL (like `https://api.nuget.org/v3/index.json`) instead of a V2 package source URL (like `https://nuget.org/`) to publish symbol packages. The NuGet protocol can be awfully confusing 😅

**What should I put into `-Source` parameter if I want to push a Symbol Package?**
I build BaGet from the source code and deploy to IIS (which is running on my company server).
The BaGet site is now serving on a sub-domain: `nuget.mycompany.com`.
I have been pushing the .nupkg file using the following command: `nuget push -Source http://nuget.mycompany.com/v3/index.json MyPackage.1.0.0.nupkg -ApiKey api-key-set-in-appsettings.json-file` and it worked without problems.

**Comment by moses70** [comment]
Folks,

reading this thread, it seems I have a similar issue;
I push several package (with .snuget) to my BaGet server without problem. Only one of them fails with:
```
PUT http://localhost:5000/api/v2/symbol/
  BadRequest http://localhost:5000/api/v2/symbol/ 2038 ms
error: Response status code does not indicate success: 400 (Bad Request).
```

It turns out that this one package contains a C++/CLI project and the pdb from it. I can fix it easily by removing the offending .pdb from the .snuget before uploading.

So the problem must be that the one .pdb is not in portable format!

The server outputs an exception which also points to this:
```
Unable to index symbol package due to exception
System.BadImageFormatException: Invalid COR20 header signature.
```

Is it actually the same what the issue starter encountered? Then my information can help.
And, of course, I would be really glad about a suggestion on how to fix this. (Skip the non-portable .pdb or make it acceptable)
